scripts/update_xspec_functions.py

```diff
@@ -36,68 +36,450 @@
 
-def print_model(model: xspec.ModelDefinition,
-                prev: Optional[str]) -> Optional[str]:
-    """Print out a model"""
+# Store a set of #ifdef...#else lines by
+# - a dictionary where the key is each "#" line and the
+#   values are the lines guarded by that check
+# - a checksum value which just gives a combination of
+#   all the "#.." lines so we can easily look for
+#   a compatible set of lines.
+#
+Checksum = str
+Storage = tuple[dict[str, list[str]], Checksum]
+
+NO_CHECKSUM: Checksum = "<none>"
+
+START_MACRO = "// Start model definitions"
+END_MACRO = "// End model definitions"
 
-    try:
-        code = xspec.model_to_compiled(model)
-    except ValueError:
-        # Assume an unsupported model
-        # sys.stderr.write(f"# Unsupported: {model.name} / {model.funcname}\n")
-        return prev
 
-    # For the moment models that are calculated per-spectrum are unsupported.
+def convert_version(smajor: str, sminor: str, smicro: str) -> Version:
+    """Convert version tuple"""
+
     try:
-        if model.flags[1] == 1:
-            return prev
-    except IndexError:
-        pass
+        major = int(smajor)
+        minor = int(sminor)
+        micro = int(smicro)
+    except ValueError as ve:
+        raise ValueError(f"version={smajor}/{sminor}/{smicro} elements must be integers") from ve
+
+    return major, minor, micro
+
+
+def parse_version(ver: str) -> Version:
+    """Convert version to its components."""
+
+    match ver.split("."):
+        case [smajor, sminor, smicro]:
+            return convert_version(smajor, sminor, smicro)
+
+        case _:
+            raise ValueError(f"version={ver} should match 'X.Y.Z'")
+
+
+def grab_version(inline: str) -> Version:
+    """Given an #ifdef/#else line, grab the version"""
+
+    toks = inline.split()
+    if len(toks) != 2:
+        raise ValueError(f"Expected two tokens in '{inline}'")
+
+    if not toks[1].startswith("XSPEC_"):
+        raise ValueError(f"Expected XSPEC_X_Y_Z in '{inline}'")
+
+    version = toks[1][6:]
+    match version.split("_"):
+        case [smajor, sminor, smicro]:
+            return convert_version(smajor, sminor, smicro)
+
+        case _:
+            raise ValueError(f"version={toks[1]} should match 'XSPEC_X_Y_Z'")
+
+
+def make_xspec_macro(version: Version) -> str:
+    """Create the XSPEC macro for this version."""
+    return "XSPEC_{0}_{1}_{2}".format(*version)
+
+
+def find_file() -> Path:
+    """Find the _xspec.cc file"""
+
+    basepath = Path("sherpa") / "astro" / "xspec" / "src" / "_xspec.cc"
+
+    # Is this run from the base directory?
+    if basepath.exists():
+        return basepath
+
+    # Is it run from scripts?
+    inpath = Path("..") / basepath
+    if inpath.exists():
+        return inpath
+
+    # Can we find it from the installed location (which implies an
+    # editable install)? This is not ideal, since it is somewhat
+    # mis-using the importlib machinery. The files is labelled as
```

Comment:
The files is labeled -> The file (singular)

Reply:
Actually, it should be "The files call is labelled as" as I'm meant to be talking about `resources.files` here. An amendment will be made.